storage/test/merkleTreeAdapter_test.cpp

```diff
@@ -944,6 +957,157 @@ TEST_P(db_adapter, no_blocks) {
   ASSERT_TRUE(value.empty());
 }
 
+TEST_P(db_adapter, state_transfer_reverse_order_with_blockchain_blocks) {
+  const auto numBlockchainBlocks = 5;
+  const auto numStBlocks = 7;
+  const auto numTotalBlocks = numBlockchainBlocks + numStBlocks;
+  const auto referenceBlockchain = createBlockchain(GetParam()->db(), numTotalBlocks);
+
+  auto adapter = DBAdapter{GetParam()->db()};
+
+  // Add blocks to the blockchain and verify both block pointers.
+  for (auto i = 1; i <= numBlockchainBlocks; ++i) {
+    ASSERT_TRUE(adapter.addLastReachableBlock(block::getData(referenceBlockchain[i - 1])).isOK());
+    ASSERT_EQ(adapter.getLatestBlock(), i);
+    ASSERT_EQ(adapter.getLastReachableBlock(), i);
+  }
+
+  // Receive more blocks from state transfer and add them to the blockchain.
+  for (auto i = numTotalBlocks; i > numBlockchainBlocks; --i) {
+    ASSERT_TRUE(adapter.addBlock(referenceBlockchain[i - 1], i).isOK());
+    ASSERT_EQ(adapter.getLatestBlock(), numTotalBlocks);
+    if (i == numBlockchainBlocks + 1) {
+      // We link the blockchain and state transfer chains at that point.
+      ASSERT_EQ(adapter.getLastReachableBlock(), numTotalBlocks);
+    } else {
+      ASSERT_EQ(adapter.getLastReachableBlock(), numBlockchainBlocks);
+    }
+
+    // Verify that initial blocks are accessible at all steps.
+    for (auto j = 1; j <= numBlockchainBlocks; ++j) {
+      auto block = Sliver{};
+      auto found = false;
+      ASSERT_TRUE(adapter.getBlockById(j, block, found).isOK());
+      ASSERT_TRUE(found);
+      const auto &referenceBlock = referenceBlockchain[j - 1];
+      ASSERT_TRUE(block == referenceBlock);
+      ASSERT_TRUE(block::getData(block) == block::getData(referenceBlock));
+      ASSERT_TRUE(block::getStateHash(block) == block::getStateHash(referenceBlock));
```

Comment:
Maybe add another check to ensure the parent hash of block is is equal to the state hash of referenceBlock-1.

Reply:
Good point. I have added a check if the parent hash of block is equal to the hash of referenceBlock - 1 . Is that the case you refer to? If not, I will add further tests in a separate PR.